File: binary tree/level_order.py
from collections import deque

class Solution(object):
    def levelOrder(self, root):
        if not root:
            return []

        result = []
        queue = deque([(root, 0)])  # each item is (node, level)

        while queue:
            node, level = queue.popleft()

            # if this is the first node of a new level, add a new list
            if len(result) <= level:
                result.append([])

            result[level].append(node.val)

            if node.left:
                queue.append((node.left, level + 1))
            if node.right:
                queue.append((node.right, level + 1))

        return result
# breadth first 
# time complexity: O(n)

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        array = []
        q= collections.deque()
        q.append(root)
        while q:
            qLen = len(q)
            subArray = []
            for i in range(qLen): # first iteration length is 1[1], second iteration 2[2, 3]
                node = q.popleft()
                subArray.append(node.val)
                if node.left: q.append(node.left)# for every iteration, you enqueue the nodes in each level so next while loop iteration size == # of nodes endqued
                if node.right: q.append(node.right)
            if subArray:
                array.append(subArray# after each for loop iteration, you add the subArray to the larger array
        return array

File: binary tree/level_order.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
  
vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> result; // Store levels

    if (root == nullptr)
    {
        return result;
    }
    // Create queue to store nodes to be processed
    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()){
        int levelSize = q.size();
        vector <int> levelNodes;

        for (int i = 0; i < levelSize; i++)
        {
            // Get current node to process
            TreeNode* currentNode = q.front();
            // Remove current node to process next node for next iteration
            q.pop();
            // Place current node value to vector storing each level
            levelNodes.push_back(currentNode->val);

            //check for left and right children if exists
            if (currentNode->left != nullptr)
            {
                q.push(currentNode->left);
            }
            if (currentNode->right != nullptr)
            {
                q.push(currentNode->right);
            }
        }
        // Store each level to the final result vector
        result.push_back(levelNodes);

    }
    return result;
}
};
